adapters/rubicon/rubicon.go

```diff
@@ -909,22 +921,8 @@ func resolveBidFloorAttributes(bidFloor float64, bidFloorCur string) (float64, s
 }
 
 func updateUserExtWithIabAttribute(userExtRP *rubiconUserExt, data []openrtb2.Data) error {
-	var segmentIdsToCopy = make([]string, 0)
-
-	for _, dataRecord := range data {
-		if dataRecord.Ext != nil {
-			var dataExtObject rubiconUserDataExt
-			err := json.Unmarshal(dataRecord.Ext, &dataExtObject)
-			if err != nil {
-				continue
-			}
-			if strings.EqualFold(dataExtObject.TaxonomyName, "iab") {
-				for _, segment := range dataRecord.Segment {
-					segmentIdsToCopy = append(segmentIdsToCopy, segment.ID)
-				}
-			}
-		}
-	}
+	userSegtaxes := []int{4}
+	var segmentIdsToCopy = getSegmentIdsToCopy(data, userSegtaxes)
```

Comment:
Nitpick: With this changes, and the addition of updateSiteExtWithIabAttribute(...) function below, functions `updateUserExtWithIabAttribute(...)` and `updateSiteExtWithIabAttribute(...)` look too much alike. Do you think it is feasible to maintain one less function if we create a generic?
```
func updateExtWithIabAttribute(target json.RawMessage, data []openrtb2.Data, segTaxes []int) error {
    var segmentIdsToCopy = getSegmentIdsToCopy(data, segTaxes)
    
    extRPTarget := make(map[string]interface{})
    
    if target != nil {
        if err := json.Unmarshal(target, &extRPTarget); err != nil {
            return &errortypes.BadInput{Message: err.Error()}
        }
    }
    
    extRPTarget["iab"] = segmentIdsToCopy
    
    if jsonTarget, err := json.Marshal(&extRPTarget); err != nil {
        return &errortypes.BadInput{Message: err.Error()}
    } else {
        target = jsonTarget
    }
    
    return nil
}
```

Reply:
I'm new to Go, so I think that it should be mistake, but when I use this function, and instead of 
```go
if err := updateUserExtWithIabAttribute(&userExtRP, userCopy.Data); err != nil {
```
use
```go
if err := updateExtWithIabAttribute(userExtRP.RP.Target, userCopy.Data, []int{4}); err != nil {
```
json test stops working, so what is my mistake?

Reply:
Thank you for addressing this comment. I believe we were having that issue because `json.RawMessage` is a slice, and it is better to return the modified slice like:
```
func updateExtWithIabAttribute(target json.RawMessage, data []openrtb2.Data, segTaxes []int) (json.RawMessage, error) {
    var segmentIdsToCopy = getSegmentIdsToCopy(data, segTaxes)

    extRPTarget := make(map[string]interface{})

    if target != nil {
        if err := json.Unmarshal(target, &extRPTarget); err != nil {
            return nil, &errortypes.BadInput{Message: err.Error()}
        }
    }

    extRPTarget["iab"] = segmentIdsToCopy

    jsonTarget, err := json.Marshal(&extRPTarget)
    if err != nil {
        return nil, &errortypes.BadInput{Message: err.Error()}
    }
    return jsonTarget, nil
}
```
Which in turn, implies the following changes:
```
757         if request.User != nil {
758             userCopy := *request.User
759 -           userExtRP := rubiconUserExt{RP: rubiconUserExtRP{Target: rubiconExt.Visitor}}
760
761 -           if err := updateUserExtWithIabAttribute(&userExtRP, userCopy.Data); err != nil {
762 +           target, err := updateExtWithIabAttribute(rubiconExt.Visitor, userCopy.Data, []int{4})
763 +           if err != nil {
764                 errs = append(errs, err)
765                 continue
766             }
767 +           userExtRP := rubiconUserExt{RP: rubiconUserExtRP{Target: target}}
                 .
                 .
                 .
852         if request.Site != nil {
853             siteCopy := *request.Site
854             siteExtRP := rubiconSiteExt{RP: rubiconSiteExtRP{SiteID: rubiconExt.SiteId}}
855             siteExtRP := rubiconSiteExt{RP: rubiconSiteExtRP{SiteID: rubiconExt.SiteId}}
856 -           if err := updateSiteExtWithIabAttribute(&siteExtRP, siteCopy.Content.Data); err != nil {
857 +           target, err := updateExtWithIabAttribute(nil, siteCopy.Content.Data, []int{1, 2})
858 +           if err != nil {
859                 errs = append(errs, err)
860                 continue
861             }
862 +           siteExtRP.RP.Target = target
863
864             siteCopy.Ext, err = json.Marshal(&siteExtRP)
adapters/rubicon/rubicon.go
```
As well as removing functions `updateUserExtWithIabAttribute` and `updateSiteExtWithIabAttribute` that, if applied this change, they won't be necessary anymore.

I agree with the changes made in this PR and will approve even if this last change is not made. Let me know whether or not you find it convenient to apply in this PR or to maybe address this later.

Reply:
Thanks, so much)))